File: src/services/decisionOptimization.ts
/** API helpers for Decision Optimization Engine. */
import { apiGet, apiPost, type ApiRequestInit } from './apiClient';
import type { DecisionStatus, DecisionWorkspaceData } from '../components/decision/types';

export async function fetchDecisionStatus(init?: ApiRequestInit): Promise<DecisionStatus> {
  return apiGet<DecisionStatus>('/api/v1/decision/status', init);
}

export async function recommendDecisions(
  body?: {
    region_id?: string;
    reservoir_id?: string;
    horizon_days?: number;
    max_actions?: number;
  },
  init?: ApiRequestInit
): Promise<{ success: boolean; message: string; data: DecisionWorkspaceData }> {
  return apiPost(
    '/api/v1/decision/recommend',
    {
      region_id: body?.region_id,
      reservoir_id: body?.reservoir_id,
      horizon_days: body?.horizon_days ?? 30,
      max_actions: body?.max_actions ?? 12,
      include_upstream: true,
    },
    init
  );
}

export async function compareStrategies(body?: {
  region_id?: string;
  reservoir_id?: string;
  horizon_days?: number;
  max_actions?: number;
  aggressiveness?: number;
}): Promise<{ success: boolean; message: string; data: DecisionWorkspaceData }> {
  return apiPost('/api/v1/decision/compare', {
    region_id: body?.region_id,
    reservoir_id: body?.reservoir_id,
    horizon_days: body?.horizon_days ?? 30,
    max_actions: body?.max_actions ?? 8,
    aggressiveness: body?.aggressiveness ?? 1.0,
  });
}

File: fastapi_app/routers/decision_optimization.py
"""HTTP API for AI Decision Optimization Engine."""
from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, status

from fastapi_app.core.dependencies import ModelRegistryDep
from fastapi_app.core.security import get_current_user
from fastapi_app.decision_engine.optimization.schemas import (
    DecisionCompareRequest,
    DecisionCompareResponse,
    DecisionRecommendRequest,
    DecisionRecommendResponse,
    DecisionStatusResponse,
)
from fastapi_app.decision_engine.optimization.service import DecisionOptimizationService

router = APIRouter(
    prefix="/api/v1/decision",
    tags=["Decision Optimization"],
    dependencies=[Depends(get_current_user)],
)


def get_decision_optimization_service(
    registry: ModelRegistryDep,
) -> DecisionOptimizationService:
    return DecisionOptimizationService(registry)


DecisionOptDep = Annotated[
    DecisionOptimizationService, Depends(get_decision_optimization_service)
]


@router.get(
    "/status",
    response_model=DecisionStatusResponse,
    summary="Decision optimization readiness and upstream model status",
)
async def decision_status(service: DecisionOptDep):
    data = service.status()
    return DecisionStatusResponse(
        ready=data["ready"],
        module=data["module"],
        module_version=data["module_version"],
        upstream=data.get("upstream") or {},
        message=data.get("message") or "",
    )


@router.post(
    "/recommend",
    response_model=DecisionRecommendResponse,
    summary="Generate ranked operational decisions from prediction outputs",
)
async def decision_recommend(payload: DecisionRecommendRequest, service: DecisionOptDep):
    try:
        data = service.recommend(
            region_id=payload.region_id,
            reservoir_id=payload.reservoir_id,
            horizon_days=payload.horizon_days,
            max_actions=payload.max_actions,
            include_upstream=payload.include_upstream,
        )
        return DecisionRecommendResponse(
            message=data.get("message", "Recommendations generated"),
            data=data,
        )
    except Exception as exc:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Decision recommendation failed: {exc}",
        ) from exc


@router.post(
    "/compare",
    response_model=DecisionCompareResponse,
    summary="Compare current strategy vs optimized strategy",
)
async def decision_compare(payload: DecisionCompareRequest, service: DecisionOptDep):
    try:
        data = service.compare(
            region_id=payload.region_id,
            reservoir_id=payload.reservoir_id,
            horizon_days=payload.horizon_days,
            max_actions=payload.max_actions,
            aggressiveness=payload.aggressiveness,
        )
        return DecisionCompareResponse(
            message=data.get("message", "Strategy comparison generated"),
            data=data,
        )
    except Exception as exc:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Decision comparison failed: {exc}",
        ) from exc
